x/programs/examples/storage/storage.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package storage
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	smath "github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/crypto/ed25519"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+const (
+	HRP           = "simulator_key_"
+	programPrefix = 0x0
+	balancePrefix = 0x1
+
+	BalanceChunks uint16 = 1
+)
+
+var (
+	ErrInvalidBalance = errors.New("invalid balance")
+	balanceKeyPool    = sync.Pool{
+		New: func() any {
+			return make([]byte, 1+ed25519.PublicKeyLen+consts.IDLen+consts.Uint16Len)
+		},
+	}
+)
+
+func ProgramPrefixKey(id []byte, key []byte) (k []byte) {
+	k = make([]byte, consts.IDLen+1+len(key))
+	k[0] = programPrefix
+	copy(k, id[:])
+	copy(k[consts.IDLen:], (key[:]))
+	return
+}
+
+//
+// Program
+//
+
+func ProgramKey(id ids.ID) (k []byte) {
+	k = make([]byte, 1+consts.IDLen)
+	copy(k[1:], id[:])
+	return
+}
+
+// [programID] -> [programBytes]
+func GetProgram(
+	ctx context.Context,
+	db state.Immutable,
+	programID ids.ID,
+) (
+	[]byte, // program bytes
+	bool, // exists
+	error,
+) {
+	k := ProgramKey(programID)
+	v, err := db.GetValue(ctx, k)
+	if errors.Is(err, database.ErrNotFound) {
+		return nil, false, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	return v, true, nil
+}
+
+// SetProgram stores [program] at [programID]
+func SetProgram(
+	ctx context.Context,
+	mu state.Mutable,
+	programID ids.ID,
+	program []byte,
+) error {
+	k := ProgramKey(programID)
+	return mu.Insert(ctx, k, program)
+}
+
+//
+// Balance
```

Comment:
Typically this storage would not have "balance" access and that would be implemented as a native smart contract. Can you elaborate on how this works (seems to be out-of-scope of this package unless I'm missing something)?

Reply:
Unless this is a type of primitive variable type we want to expose (could buy that)?

Reply:
yeah I brought in a bunch of stuff for the example/storage that needs to get cleaned up.